wlauto/tests/test_environment.py

```diff
@@ -0,0 +1,289 @@
+        self.count += 1
+
+
+class SubSubClass(SubClass):
+
+    def __init__(self):
+        super(SubSubClass, self).__init__()
+
+    @once
+    def initilise_once(self):
+        self.count += 1
+
+    @once_per_class
+    def initilise_once_per_class(self):
+        self.count += 1
+
+    @once_per_instance
+    def initilise_once_per_instance(self):
+        self.count += 1
+
+
+class AnotherClass(object):
+
+    def __init__(self):
+        self.count = 0
+
+    @once
+    def initilise_once(self):
+        self.count += 1
+
+    @once_per_class
+    def initilise_once_per_class(self):
+        self.count += 1
+
+    @once_per_instance
+    def initilise_once_per_instance(self):
+        self.count += 1
+
+
+class EnvironmentManagementTest(TestCase):
+
+    def test_duplicate_environment(self):
+        init_environment('ENVIRONMENT')
+        assert_raises(ValueError, init_environment, 'ENVIRONMENT')
+
+    def test_reset_missing_environment(self):
+        assert_raises(ValueError, reset_environment, 'MISSING')
+
+    def test_reset_current_environment(self):
+        activate_environment('CURRENT_ENVIRONMENT')
+        t1 = TestClass()
+        t1.initilise_once()
+        assert_equal(t1.count, 1)
+
+        reset_environment()
+        t1.initilise_once()
+        assert_equal(t1.count, 2)
+
+    def test_switch_environment(self):
+        activate_environment('ENVIRONMENT1')
+        t1 = TestClass()
+        t1.initilise_once()
+        assert_equal(t1.count, 1)
+
+        activate_environment('ENVIRONMENT2 ')
+        t1.initilise_once()
+        assert_equal(t1.count, 2)
+
+    def test_reset_environment_name(self):
+        activate_environment('ENVIRONMENT')
+        t1 = TestClass()
+        t1.initilise_once()
+        assert_equal(t1.count, 1)
+
+        reset_environment('ENVIRONMENT')
+        t1.initilise_once()
+        assert_equal(t1.count, 2)
+
+
+class OnlyOnceEnvironmentTest(TestCase):
+
+    def setUp(self):
+        activate_environment('TEST_ENVIRONMENT')
+
+    def tearDown(self):
+        reset_environment('TEST_ENVIRONMENT')
+
+    def test_single_instance(self):
+        t1 = TestClass()
+        ac = AnotherClass()
+
+        t1.initilise_once()
+        assert_equal(t1.count, 1)
+
+        t1.initilise_once()
+        assert_equal(t1.count, 1)
+
+        ac.initilise_once()
+        assert_equal(ac.count, 1)
+
+
+    def test_mulitple_instances(self):
+        t1 = TestClass()
+        t2 = TestClass()
+
+        t1.initilise_once()
+        assert_equal(t1.count, 1)
+
+        t2.initilise_once()
+        assert_equal(t2.count, 0)
+
+
+    def test_sub_classes(self):
+        t1 = TestClass()
+        sc = SubClass()
+        ss = SubSubClass()
+
+        t1.initilise_once()
+        assert_equal(t1.count, 1)
+
+        sc.initilise_once()
+        sc.initilise_once()
+        assert_equal(sc.count, 0)
+
+        ss.initilise_once()
+        ss.initilise_once()
+        assert_equal(ss.count, 0)
```

Comment:
To complete this test, I would make sure that calling `t1.initialise_once()` again does not increment the count.